### PR TITLE
Fix file.replace mtime update on no changes

### LINKED ISSUE #61743: [BUG] file.replace updates the mtime for a file with no changes
**Description**
When using the `file.replace` state, the file's modification date is changed even if there are no content changes.

**Setup**

Create this SLS file and run it once to create the contents.

```
# test.sls

empty_file:
  file.managed:
    - name: /tmp/myfile
    - replace: False

replace_test:
  file.replace:
    - name: /tmp/myfile
    - prepend_if_not_found: True
    - pattern: "^# salt"
    - repl: "# salt"
```
  
**Steps to Reproduce the behavior**

Subsequent runs of the state above will always update the modified time of the file.

```
root@d10-salt-01:~# date
Tue Mar  1 13:58:13 UTC 2022

root@d10-salt-01:~# ll /tmp/myfile
-rw-r--r-- 1 root root 7 Mar  1 13:57 /tmp/myfile

root@d10-salt-01:~# salt-call state.apply replace --state-output=terse_id
local:
  Name: empty_file - Function: file.managed - Result: Clean Started: - 13:58:39.627202 Duration: 2.488 ms
  Name: replace_test - Function: file.replace - Result: Clean Started: - 13:58:39.629764 Duration: 2.308 ms

Summary for local
------------
Succeeded: 2
Failed:    0
------------
Total states run:     2
Total run time:   4.796 ms

root@d10-salt-01:~# ll /tmp/myfile
-rw-r--r-- 1 root root 7 Mar  1 13:58 /tmp/myfile
```

**Expected behavior**
If the contents aren't updated, the mtime stamp shouldn't be updated.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```yaml
Salt Version:
          Salt: 3004
 
Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.5
     gitpython: 2.1.11
        Jinja2: 2.10
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.7.3 (default, Jan 22 2021, 20:04:44)
  python-gnupg: Not Installed
        PyYAML: 3.13
         PyZMQ: 17.1.2
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.1
 
System Versions:
          dist: debian 10 buster
        locale: UTF-8
       machine: x86_64
       release: 4.19.0-18-amd64
        system: Linux
       version: Debian GNU/Linux 10 buster
```
</details>

**Additional context**
This was previously surfaced in issue #55297, addressed in PR #56359, and then (inadvertently?) reverted in PR #56621

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
This PR reintroduces the fix from #56359 which was reverted (inadvertently?) in #56621

### What issues does this PR fix or reference?
Fixes: #61743

### Previous Behavior
File mtime always updated when the pattern doesn't match the replacement value.

### New Behavior
File mtime only updated when the content changes.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by nicholasmhughes** [comment]
re-run pr-debian-9-amd64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by Ch3LL** [comment]
@waynew can you review since you were the author of https://github.com/saltstack/salt/pull/56621

File: salt/modules/file.py [hunk 1]
```diff
@@ -2593,6 +2593,8 @@ def replace(
                     else r_data.splitlines(True)
                 )
                 new_file = result.splitlines(True)
+                if orig_file == new_file:
+                    has_changes = False
```
Comment (waynew) [inline salt/modules/file.py:2597]: **suggestion (non-blocking)** this could just be

```suggestion
                has_changes = orig_file != new_file
```

**Review by waynew — APPROVED** [review]
Good find! I went back to double check, and it looks like that was an inadvertent change, and lacking test coverage it wasn't caught :upside_down_face: 

Just to be on the safe side I pulled this down and banged on it a bit. I couldn't make things break without *actually* breaking things. This looks like a good fix - it *could* be spelled differently but I think that's just a matter of taste.

:+1:

**Pull request merged**